sdk/src/azure/core/az_json_parser.c

```diff
@@ -1,558 +1,693 @@
+  json_parser->_internal.bytes_consumed += az_span_size(remaining) - az_span_size(json);
+
+  return json;
 }
 
-AZ_NODISCARD AZ_INLINE az_result
-az_json_parser_push_stack(az_json_parser* ref_json_parser, _az_json_stack json_stack)
+AZ_NODISCARD static az_result _az_json_parser_process_container_end(
+    az_json_parser* json_parser,
+    az_json_token_kind token_kind)
 {
-  if (ref_json_parser->_internal.stack >> (_az_MAX_JSON_STACK_SIZE - 1) != 0)
+  // The JSON payload is invalid if it has a mismatched container end without a matching open.
+  if ((token_kind == AZ_JSON_TOKEN_END_OBJECT
+       && _az_json_stack_peek(&json_parser->_internal.bit_stack) != _az_JSON_STACK_OBJECT)
+      || (token_kind == AZ_JSON_TOKEN_END_ARRAY
+          && _az_json_stack_peek(&json_parser->_internal.bit_stack) != _az_JSON_STACK_ARRAY))
   {
-    return AZ_ERROR_JSON_NESTING_OVERFLOW;
+    return AZ_ERROR_PARSER_UNEXPECTED_CHAR;
   }
-  ref_json_parser->_internal.stack = (ref_json_parser->_internal.stack << 1) | json_stack;
+
+  az_span token = _get_remaining_json(json_parser);
+  _az_json_stack_pop(&json_parser->_internal.bit_stack);
+  _az_json_parser_update_state(json_parser, token_kind, az_span_slice(token, 0, 1), 1);
   return AZ_OK;
 }
 
-AZ_NODISCARD AZ_INLINE az_result az_json_stack_pop(_az_json_stack* json_stack)
+AZ_NODISCARD static az_result _az_json_parser_process_container_start(
+    az_json_parser* json_parser,
+    az_json_token_kind token_kind,
+    _az_json_stack_item container_kind)
 {
-  if (*json_stack <= 1)
+  // The current depth is equal to or larger than the maximum allowed depth of 64. Cannot read the
+  // next JSON object or array.
+  if (json_parser->_internal.bit_stack._internal.current_depth >= _az_MAX_JSON_STACK_SIZE)
   {
-    return AZ_ERROR_JSON_INVALID_STATE;
+    return AZ_ERROR_JSON_NESTING_OVERFLOW;
   }
-  *json_stack >>= 1;
+
+  az_span token = _get_remaining_json(json_parser);
+
+  _az_json_stack_push(&json_parser->_internal.bit_stack, container_kind);
+  _az_json_parser_update_state(json_parser, token_kind, az_span_slice(token, 0, 1), 1);
   return AZ_OK;
 }
 
-AZ_NODISCARD AZ_INLINE az_result az_json_parser_pop_stack(az_json_parser* json_parser)
+AZ_NODISCARD static bool _az_is_valid_escaped_character(uint8_t byte)
 {
-  return az_json_stack_pop(&json_parser->_internal.stack);
+  switch (byte)
+  {
+    case '\\':
+    case '"':
+    case '/':
+    case 'b':
+    case 'f':
+    case 'n':
+    case 'r':
+    case 't':
+      return true;
+    default:
+      return false;
+  }
 }
 
-AZ_NODISCARD az_result az_json_parser_init(az_json_parser* json_parser, az_span json_buffer)
+// A hex digit is valid if it is in the range: [0..9] | [A..F] | [a..f]
+// Otherwise, return false.
+AZ_NODISCARD static bool _az_is_hex_digit(uint8_t byte)
 {
-  *json_parser = (az_json_parser){ ._internal = { .reader = json_buffer, .stack = 1 } };
-  return AZ_OK;
+  return (byte >= '0' && byte <= '9') || (byte >= 'a' && byte <= 'f')
```

Comment:
```
#include <ctype.h>
isxdigit()
```
https://en.cppreference.com/w/c/string/byte/isxdigit

Reply:
unlike some other functions, such as isalnum(), this one is not affected by locale, so it is safe to use and expect it works the same on all locales.